### DISCUSSION TITLE
Is it possible to override theme colors with arbitrary, non-tokenized color values?

### DISCUSSION THREAD

**Discussion by fwextensions** [uswds/uswds#4859] [Q&A] [answered]
[This PR](https://github.com/uswds/uswds/pull/3258) seems to imply that the capability to customize colors was added in 2.4.0.  But using 2.13.3, trying to override a color variable using something like `$theme-color-primary: #425CC7;` throws an error and refuses to compile:

```
Error: "Only use quoted color tokens in USWDS functions and mixins. See designsystem.digital.gov/design-tokens/color for more information."
     ╷
1364 │   $value: smart-quote($value);
     │           ^^^^^^^^^^^^^^^^^^^
     ╵
```

Is using arbitrary colors actually supported?

**Comment by amyleadem** [comment]
Hi @fwextensions,
Do you have any code that runs the customized variable through the `color()` function (or any USWDS function/mixin)? If so, that can cause the error you are seeing. 

For example, I receive the error you listed above when I run the customized theme variable through the color() function, like this:
```
.test {
    background-color: color($theme-color-primary);
}
```
But it compiles successfully when I directly define the value like this:
```
.test {
    background-color: $theme-color-primary;
}
```

I did not receive any errors when I customized `$theme-color-primary` to `#425CC7` with the base USWDS code, so I suspect there is custom code in your project that might be attempting to pass something other than a quoted color token into a USWDS function/mixin.

I hope this helps. Please let me know if you still have questions. Thanks!

**Reply by fwextensions** [reply]
Yes, the `color()` function is used all over, as it is in the USWDS source as well.  I assumed that was just how colors defined by the theme variables are converted to CSS values.  If you have to know how a color was defined before you can use it, then that's not a very usable theming system.  And it would fail with all of the built-in USWDS mixins and functions, e.g., `add-link-styles()`.  

Surely there's some way to use arbitrary color values with the `color()` function?  Can the color token strings themselves be overridden?

**Reply by mejiaj** [reply]
@fwextensions sorry to hear you're running into issues. Can you provide a code example or a link to the repo so we can take a look? The `color()` function _should_ be able to take a hex value as long as it isn't quoted.

**Reply by fwextensions** [reply]
It's just a one-line change.  All I'm doing is editing the `_uswds-theme-color.scss` file in the project, which is copied from the defaults.  Below is the primary colors section of the file, where the `blue-60v` token is getting replaced with a normal CSS hex value.  That triggers the "color tokens" error described above.

```scss
// Primary colors
$theme-color-primary-family: 'blue';
$theme-color-primary-lightest: false;
$theme-color-primary-lighter: 'blue-10';
$theme-color-primary-light: 'blue-30';
$theme-color-primary: #425CC7;
//$theme-color-primary: 'blue-60v';
$theme-color-primary-vivid: 'blue-warm-60v';
$theme-color-primary-dark: 'blue-warm-70v';
$theme-color-primary-darker: 'blue-warm-80v';
$theme-color-primary-darkest: false;
```

**Reply by mejiaj** [reply]
@fwextensions I was able to replicate that error and I think there's some confusion on how these are used.

For example, passing the variable in the color function wouldn't work because it expects a token. But if you pass your theme colors as tokens, for example `color("primary")` instead of the variable it will work.

```diff
// In settings
$theme-color-primary: #18f;

// In custom styles
.custom-selector {
- // ❌ Doesn't work
-  color: color($theme-color-primary);
+ // ✅ Works 
+ color: color("primary");
}
```

**Reply by fwextensions** [reply]
Thanks for the investigation!  This code base uses the `color($theme-color-primary)` approach everywhere, so I was just following that pattern.  It's confusing because passing a color variable works out of the box, until you change it to a regular color string.  And then using the bare variable, like `color: $theme-color-primary` *would* work, but using bare variables you hadn't customized wouldn't.  

It's also confusing because the USWDS source code doesn't consistently use theme color strings:

```scss
.usa-banner {
  @include typeset($theme-banner-font-family);
  @include border-box-sizing;
  background-color: color($theme-banner-background-color);
```

I guess the only solution is updating all the `color()` calls to use the token strings.

**Reply by mejiaj** [reply]
>Thanks for the investigation! This code base uses the color($theme-color-primary) approach everywhere, so I was just following that pattern.

Understandable, but if you look at component settings they use tokens. It seems more beneficial to use `color("primary")` vs `color($theme-color-primary)` because it's shorter and easier to remember anyways.

**Reply by fwextensions** [reply]
To be clear, I was referring to the code base I'm working on, not the USWDS source.  Not sure how the project I'm working on got started down this path, but possibly an older version of USWDS was using the variables in more places.

Seems like things would be clearer if the `uswds-theme-color.scss` file was set up something like:

```scss
@use ('theme-colors') with (
  'primary-family': 'blue',
  'primary': 'blue-60v',
  ...
)
```

I don't know if that's valid SCSS, but setting the colors via strings instead of variables would clarify how to use `color()`, I think.

Anyway, I searched and replaced all the `color()` calls, and the custom color is working now.  Thanks for your help!

**Comment by fwextensions** [comment, marked as answer]
As @mejiaj [says](https://github.com/uswds/uswds/discussions/4859#discussioncomment-3157496):

> if you pass your theme colors as tokens, for example `color("primary")` instead of the variable it will work.

Replacing all the calls to things like `color($theme-color-primary-dark)` with `color("primary-dark")` enabled the theme to work with colors that were customized with normal CSS values instead of [USWDS system color tokens](https://designsystem.digital.gov/design-tokens/color/system-tokens/).

**Reply by mejiaj** [reply]
Glad you got the issue fixed!

**Comment by ryparker** [comment]
The documented color token syntax resolved the error, as the author confirmed above. I'm closing this as answered; the [working solution](https://github.com/uswds/uswds/discussions/4859#discussioncomment-3175067) remains available for reference.
